File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/adl/holder/pipeline/ADLOutput.java
package com.kite9.k9server.adl.holder.pipeline;

import com.kite9.k9server.adl.holder.meta.MetaRead;
import org.kite9.diagram.logging.Kite9ProcessingException;

import com.kite9.k9server.adl.format.media.DiagramFileFormat;
import com.kite9.k9server.adl.format.media.DiagramFormat;

/**
 * Third point in the pipeline: ADL is now converted to it's output format.
 * 
 * @author robmoffat
 *
 */
public interface ADLOutput<X extends DiagramFormat> extends XMLBase, MetaRead {
	
	/**
	 * Returns the format used to create this output.
	 */
	public X getFormat();
	
	byte[] getAsBytes();
	
	/**
	 * Return as UTF-8 string, unless this is a binary format {@link DiagramFileFormat}.isBinaryFormat();
	 */
	String getAsString() throws Kite9ProcessingException;
	
	/**
	 * Creator of this.
	 */
	public ADLDom originatingADLDom();
}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/k9server/adl/holder/pipeline/ADLOutput.kt
package com.kite9.k9server.adl.holder.pipeline

import com.kite9.k9server.adl.format.media.DiagramWriteFormat
import com.kite9.k9server.adl.holder.meta.MetaRead

/**
 * Third point in the pipeline: ADL is now converted to it's output format.
 *
 * @author robmoffat
 */
interface ADLOutput<X> : XMLBase, MetaRead {

    /**
     * Returns the format used to create this output.
     */
    val format: DiagramWriteFormat<X>
    val asBytes: ByteArray

    /**
     * Return as UTF-8 string, unless this is a binary format [DiagramFileFormat].isBinaryFormat();
     */
    val asString: String

    /**
     * Returns the produced format itself.
     */
    val value: X

    /**
     * Creator of this.
     */
    val originatingADLDom: ADLDom
}